File: Mediconnet-Frontend/src/app/services/medecin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface MedecinProfileDto {
  idUser: number;
  nom: string;
  prenom: string;
  email: string;
  telephone?: string;
  adresse?: string;
  photo?: string;
  sexe?: string;
  naissance?: string;
  numeroOrdre?: string;
  specialite?: string;
  idSpecialite?: number;
  service?: string;
  idService: number;
  createdAt?: string;
}

export interface MedecinDashboardDto {
  totalPatients: number;
  consultationsMois: number;
  rdvAujourdHui: number;
  rdvAVenir: number;
  ordonnancesMois: number;
  examensMois: number;
}

export type AgendaSlotStatus = 'disponible' | 'occupe' | 'indisponible' | 'passe';

export interface AgendaSlotDto {
  dateHeure: string;
  duree: number;
  statut: AgendaSlotStatus;
  disponible: boolean;
  raison?: string;
  idRendezVous?: number;
  patientNom?: string;
  patientPrenom?: string;
  motif?: string;
}

export interface AgendaJourDto {
  date: string;
  jourSemaine: string;
  slots: AgendaSlotDto[];
  totalDisponibles: number;
  totalOccupes: number;
  totalIndisponibles: number;
}

export interface AgendaResponse {
  dateDebut: string;
  dateFin: string;
  jours: AgendaJourDto[];
  prochainRdv?: AgendaSlotDto;
}

export interface RendezVousMedecinDto {
  idConsultation: number;
  idRendezVous: number;
  dateHeure: string;
  duree: number;
  statut: string;
  motif?: string;
  typeRdv: string;
  dateCreation?: string;
  dateModification?: string | null;
  patientNom: string;
  patientPrenom: string;
  patientId: number;
  isPremiereConsultation?: boolean;
  specialiteId?: number;
  origine?: 'accueil' | 'rdv_confirme';
  heureArrivee?: string | null;
}

export interface UpdateMedecinProfileRequest {
  telephone?: string;
  adresse?: string;
}

@Injectable({
  providedIn: 'root'
})
export class MedecinService {
  private apiUrl = `${environment.apiUrl}/medecin`;

  constructor(private http: HttpClient) {}

  /**
   * Récupérer le profil complet du médecin connecté
   */
  getProfile(): Observable<MedecinProfileDto> {
    return this.http.get<MedecinProfileDto>(`${this.apiUrl}/profile`);
  }

  /**
   * Récupérer les statistiques du dashboard médecin
   */
  getDashboard(): Observable<MedecinDashboardDto> {
    return this.http.get<MedecinDashboardDto>(`${this.apiUrl}/dashboard`);
  }

  /**
   * Mettre à jour le profil du médecin
   */
  updateProfile(data: UpdateMedecinProfileRequest): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(`${this.apiUrl}/profile`, data);
  }

  /**
   * Récupérer l'agenda du médecin (créneaux et RDV)
   */
  getAgenda(dateDebut: string, dateFin: string): Observable<AgendaResponse> {
    return this.http.get<AgendaResponse>(`${this.apiUrl}/agenda`, {
      params: { dateDebut, dateFin }
    });
  }

  /**
   * Récupérer les rendez-vous du jour
   */
  getRdvAujourdHui(): Observable<RendezVousMedecinDto[]> {
    return this.http.get<RendezVousMedecinDto[]>(`${this.apiUrl}/rdv/aujourdhui`);
  }

  /**
   * Récupérer les prochains rendez-vous
   */
  getProchainRdv(limite: number = 5): Observable<RendezVousMedecinDto[]> {
    return this.http.get<RendezVousMedecinDto[]>(`${this.apiUrl}/rdv/prochains`, {
      params: { limite: limite.toString() }
    });
  }
}

File: Mediconnet-Backend/Services/MedecinService.cs
using Mediconnet_Backend.Core.Interfaces.Services;
using Mediconnet_Backend.Data;
using Mediconnet_Backend.DTOs.Medecin;
using Microsoft.EntityFrameworkCore;

namespace Mediconnet_Backend.Services;

/// <summary>
/// Service pour la gestion des médecins
/// Contient la logique métier liée aux médecins
/// </summary>
public class MedecinService : IMedecinService
{
    private readonly ApplicationDbContext _context;
    private readonly ILogger<MedecinService> _logger;

    public MedecinService(ApplicationDbContext context, ILogger<MedecinService> logger)
    {
        _context = context;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<MedecinProfileDto?> GetProfileAsync(int userId)
    {
        var medecin = await _context.Medecins
            .Include(m => m.Utilisateur)
            .Include(m => m.Service)
            .FirstOrDefaultAsync(m => m.IdUser == userId);

        if (medecin?.Utilisateur == null)
            return null;

        // Récupérer la spécialité
        string? specialiteNom = null;
        if (medecin.IdSpecialite.HasValue)
        {
            var specialite = await _context.Specialites
                .FirstOrDefaultAsync(s => s.IdSpecialite == medecin.IdSpecialite.Value);
            specialiteNom = specialite?.NomSpecialite;
        }

        return new MedecinProfileDto
        {
            IdUser = medecin.IdUser,
            Nom = medecin.Utilisateur.Nom,
            Prenom = medecin.Utilisateur.Prenom,
            Email = medecin.Utilisateur.Email,
            Telephone = medecin.Utilisateur.Telephone,
            Adresse = medecin.Utilisateur.Adresse,
            Photo = medecin.Utilisateur.Photo,
            Sexe = medecin.Utilisateur.Sexe,
            Naissance = medecin.Utilisateur.Naissance,
            NumeroOrdre = medecin.NumeroOrdre,
            Specialite = specialiteNom,
            IdSpecialite = medecin.IdSpecialite,
            Service = medecin.Service?.NomService,
            IdService = medecin.IdService,
            CreatedAt = medecin.Utilisateur.CreatedAt
        };
    }

    /// <inheritdoc />
    public async Task<MedecinDashboardDto> GetDashboardAsync(int userId)
    {
        var now = DateTime.Now;
        var debutMois = new DateTime(now.Year, now.Month, 1);
        var finMois = debutMois.AddMonths(1);

        // Nombre de patients distincts
        var totalPatients = await _context.RendezVous
            .Where(r => r.IdMedecin == userId && r.Statut != "annule")
            .Select(r => r.IdPatient)
            .Distinct()
            .CountAsync();

        // Consultations ce mois
        var consultationsMois = await _context.RendezVous
            .CountAsync(r => r.IdMedecin == userId && 
                       r.DateHeure >= debutMois && 
                       r.DateHeure < finMois &&
                       r.Statut == "termine");

        // RDV aujourd'hui
        var rdvAujourdHui = await _context.RendezVous
            .CountAsync(r => r.IdMedecin == userId && 
                       r.DateHeure.Date == now.Date &&
                       r.Statut != "annule");

        // RDV à venir
        var rdvAVenir = await _context.RendezVous
            .CountAsync(r => r.IdMedecin == userId && 
                       r.DateHeure > now &&
                       (r.Statut == "planifie" || r.Statut == "confirme"));

        return new MedecinDashboardDto
        {
            TotalPatients = totalPatients,
            ConsultationsMois = consultationsMois,
            RdvAujourdHui = rdvAujourdHui,
            RdvAVenir = rdvAVenir,
            OrdonnancesMois = 0,
            ExamensMois = 0
        };
    }

    /// <inheritdoc />
    public async Task<bool> UpdateProfileAsync(int userId, UpdateMedecinProfileRequest request)
    {
        var medecin = await _context.Medecins
            .Include(m => m.Utilisateur)
            .FirstOrDefaultAsync(m => m.IdUser == userId);

        if (medecin?.Utilisateur == null)
            return false;

        if (!string.IsNullOrEmpty(request.Telephone))
            medecin.Utilisateur.Telephone = request.Telephone;
        if (!string.IsNullOrEmpty(request.Adresse))
            medecin.Utilisateur.Adresse = request.Adresse;

        medecin.Utilisateur.UpdatedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Profile updated for medecin {UserId}", userId);
        return true;
    }

    /// <inheritdoc />
    public async Task<MedecinAgendaDto> GetAgendaAsync(int userId, DateTime dateDebut, DateTime dateFin)
    {
        // Récupérer les créneaux configurés
        var creneaux = await _context.CreneauxDisponibles
            .Where(c => c.IdMedecin == userId && c.Actif)
            .Select(c => new CreneauAgendaDto
            {
                Id = c.IdCreneau,
                JourSemaine = c.JourSemaine.ToString(),
                HeureDebut = c.HeureDebut,
                HeureFin = c.HeureFin,
                DureeMinutes = c.DureeParDefaut,
                Actif = c.Actif
            })
            .ToListAsync();

        // Récupérer les RDV existants
        var rdvs = await _context.RendezVous
            .Include(r => r.Patient)
                .ThenInclude(p => p!.Utilisateur)
            .Where(r => r.IdMedecin == userId && 
                       r.DateHeure >= dateDebut && 
                       r.DateHeure <= dateFin &&
                       r.Statut != "annule")
            .Select(r => new RdvAgendaDto
            {
                Id = r.IdRendezVous,
                DateHeure = r.DateHeure,
                Statut = r.Statut,
                PatientNom = r.Patient != null && r.Patient.Utilisateur != null 
                    ? $"{r.Patient.Utilisateur.Prenom} {r.Patient.Utilisateur.Nom}" 
                    : "Patient inconnu",
                Motif = r.Motif
            })
            .ToListAsync();

        return new MedecinAgendaDto
        {
            Creneaux = creneaux,
            RendezVous = rdvs
        };
    }
}
